rsmc
====

RNAseq Somatic Mutation Caller

====

Build from source

1) Install Visual Studio 2010

2) Install 7-zip to "C:\Program Files\7-Zip\"

3) download following source codes to same directory:

git://github.com:shengqh/RCPA.Core.git

git://github.com:shengqh/RCPA.Proteomics.git

git://github.com:shengqh/CQS.Core.git

git://github.com:shengqh/SQH.Lib.git

git://github.com/shengqh/rsmc.git

4) open rsmc/RSMC.sln at VisualStudio 2010 and build.

5) Run DeployZip function in Deploy.cs using NUnit test. You will get RSMC.X.X.X.X.zip in your Debug/Release directory.

====

Download binary file

https://sourceforge.net/projects/rsmc/files/latest/download?source=files

====

Requirement

1) mono for running program on linux

2) samtools for pileup

3) R for filter

4) annovar for annotation

====

Configuration

A configuration file rsmc.linux (or rsmc.win) has been included or will be generated automatically.

<?xml version="1.0" encoding="utf-8"?>
<config>
  <program name="annovar" command="summarize_annovar.pl">
    <parameterset name="hg19">
      <parameter name="verdbsnp" default="137" />
      <parameter name="ver1000g" default="1000g2012apr" />
      <parameter name="veresp" default="6500si" />
      <parameter name="genetype" default="refgene" />
      <parameter name="databasedir" default="/scratch/cqs/shengq1/references/annovar/humandb/" />
    </parameterset>
  </program>
</config>

====

Usage

RNAseq Somatic Mutation Caller (RSMC) 1.0.0.0

Copyright (C) 2013 Center for Quantitative Sciences/VUMC

  help          Display more information on a specific command.

  pileup        Initialize candidates from samtools mpileup result.

  filter        Filter candidates by logistic regression model.

  annotation    Annotate mutation using varies tools.

  all           pileup/filter/annotate data

1) pileup

  -t STRING, --type=STRING            Required. Where to read/generate mpileup
                                      result file (bam/mpileup/console)

  -b BAM1,BAM2, --bam=BAM1,BAM2       Bam files used in samtools mpileup (INPUT
                                      MUST BE normal_bam,tumor_bam), only two
                                      bam files are allowed and the second one
                                      will be assumed as from tumor sample
                                      which contains somatic mutation

  -f FILE, --fasta=FILE               Genome fasta file for samtools mpileup

  -r STRING, --chromosomes=STRING     Default(all chromosomes in genome fasta
                                      file) Chromosome names (separted by ',')

  -n INT, --read_quality=INT          (Default: 20) Minimum mapQ of read for
                                      samtools mpileup

  -m FILE, --mpileup=FILE             Samtools mpileup result file

  -e DOUBLE, --pvalue=DOUBLE          (Default: 0.01) pvalue used for
                                      significance test

  -q INT, --base_quality=INT          (Default: 20) Minimum base quality for
                                      mpileup result filter

  -g DOUBLE, --percentage=DOUBLE      (Default: 0.1) Minimum percentage of
                                      minor allele at at tumor sample (second
                                      sample)

  -d INT, --read_depth=INT            (Default: 10) Minimum read depth of base
                                      passed mapping quality filter in each
                                      sample

  -p, --filter_position               (Default: true) Filter mpileup result
                                      based on position bias by fisher exact
                                      test

  -s, --filter_strand                 (Default: true) Filter mpileup result
                                      based on strand bias by fisher exact test

  -c INT, --thread_count=INT          (Default: 1) Number of thread

  -o DIRECTORY, --output=DIRECTORY    Required. Output directory

2) filter

  -v DOUBLE, --pvalue=DOUBLE                 (Default: 0.01) pvalue used for
                                             significance test

  -q INT, --base_quality=INT                 (Default: 20) Minimum base quality
                                             for mpileup result filter

  -o FILE, --output=FILE                     Required. Output file

  -d DIRECTORY, --candidate_dir=DIRECTORY    Required. Directory including
                                             candidate site files

3) annotation

  -i FILE, --input=FILE            Required. Input file generated by filter
                                   function.

  --annovar                        Annotation by annovar.

  --annovar_set_default            (Default: false) Set current setting as
                                   annovar default setting

  --annovar_db=DIRECTORY           The directory contains annovar databases

  --annovar_buildver=STRING        Annovar database version (such as hg19,
                                   mm10)

  --annovar_verdbsnp=STRING        dbSNP version to use

  --annovar_ver1000g=STRING        1000G version

  --annovar_veresp=STRING          ESP version

  --annovar_genetype=STRING        (Default: refgene) gene definition can be
                                   refgene , knowngene, ensgene

  --distance                       Annotation by distance to entries from
                                   assigned bed files.

  --distance_insertion_bed=FILE    Insertion bed file for distance calculation.

  --distance_deletion_bed=FILE     Deletion bed file for distance calculation.

  --distance_junction_bed=FILE     Junction bed file for distance calculation.

  --rnaediting                     Annotation by rna editing database.

  --rnaediting_db=FILE             The rna editing database file

4) all (the combination of pileup and annotation)

  --annovar                           Annotation by annovar.

  --annovar_set_default               (Default: false) Set current setting as
                                      annovar default setting

  --annovar_db=DIRECTORY              The directory contains annovar databases

  --annovar_buildver=STRING           Annovar database version (such as hg19,
                                      mm10)

  --annovar_verdbsnp=STRING           dbSNP version to use

  --annovar_ver1000g=STRING           1000G version

  --annovar_veresp=STRING             ESP version

  --annovar_genetype=STRING           (Default: refgene) gene definition can be
                                      refgene , knowngene, ensgene

  --distance                          Annotation by distance to entries from
                                      assigned bed files.

  --distance_insertion_bed=FILE       Insertion bed file for distance
                                      calculation.

  --distance_deletion_bed=FILE        Deletion bed file for distance
                                      calculation.

  --distance_junction_bed=FILE        Junction bed file for distance
                                      calculation.

  --rnaediting                        Annotation by rna editing database.

  --rnaediting_db=FILE                The rna editing database file

  -t STRING, --type=STRING            Required. Where to read/generate mpileup
                                      result file (bam/mpileup/console)

  -b BAM1,BAM2, --bam=BAM1,BAM2       Bam files used in samtools mpileup (INPUT
                                      MUST BE normal_bam,tumor_bam), only two
                                      bam files are allowed and the second one
                                      will be assumed as from tumor sample
                                      which contains somatic mutation

  -f FILE, --fasta=FILE               Genome fasta file for samtools mpileup

  -r STRING, --chromosomes=STRING     Default(all chromosomes in genome fasta
                                      file) Chromosome names (separted by ',')

  -n INT, --read_quality=INT          (Default: 20) Minimum mapQ of read for
                                      samtools mpileup

  -m FILE, --mpileup=FILE             Samtools mpileup result file

  -e DOUBLE, --pvalue=DOUBLE          (Default: 0.01) pvalue used for
                                      significance test

  -q INT, --base_quality=INT          (Default: 20) Minimum base quality for
                                      mpileup result filter

  -g DOUBLE, --percentage=DOUBLE      (Default: 0.1) Minimum percentage of
                                      minor allele at at tumor sample (second
                                      sample)

  -d INT, --read_depth=INT            (Default: 10) Minimum read depth of base
                                      passed mapping quality filter in each
                                      sample

  -p, --filter_position               (Default: true) Filter mpileup result
                                      based on position bias by fisher exact
                                      test

  -s, --filter_strand                 (Default: true) Filter mpileup result
                                      based on strand bias by fisher exact test

  -c INT, --thread_count=INT          (Default: 1) Number of thread

  -o DIRECTORY, --output=DIRECTORY    Required. Output directory

====

mono rsmc.exe pileup -t bam